rust/sbd-e2e-crypto-client/src/lib.rs

```diff
@@ -334,66 +105,253 @@ impl SbdClientCrypto {
         let (client, recv) =
             sbd_client::SbdClient::connect_config(url, &crypto, client_config)
                 .await?;
-        let inner = Arc::new(tokio::sync::Mutex::new(Some(Inner {
-            config,
-            crypto,
+
+        let client = Arc::new(tokio::sync::Mutex::new(client));
+        let inner = Arc::new(Mutex::new(Inner::new(config, crypto)));
+
+        let recv = MsgRecv {
+            inner: inner.clone(),
+            recv,
+            client: Arc::downgrade(&client),
+        };
+
+        let this = Self {
+            pub_key,
+            inner,
             client,
-            map: HashMap::default(),
-        })));
-        let weak_inner = Arc::downgrade(&inner);
-        Ok((
-            Self { pub_key, inner },
-            MsgRecv {
-                inner: weak_inner,
-                recv,
-            },
-        ))
+        };
+
+        Ok((this, recv))
     }
 
     /// Get the public key of this node.
     pub fn pub_key(&self) -> &PubKey {
         &self.pub_key
     }
 
+    /// Get the current active "connected" peers.
+    pub fn active_peers(&self) -> Vec<PubKey> {
+        let mut inner = self.inner.lock().unwrap();
+        let max_idle = inner.config.max_idle;
+        Inner::prune(&mut inner.c_map, max_idle);
+        inner.c_map.keys().cloned().collect()
+    }
+
     /// Assert that we are connected to a peer without sending any data.
     pub async fn assert(&self, pk: &PubKey) -> Result<()> {
-        let mut lock = self.inner.lock().await;
-        if let Some(inner) = &mut *lock {
-            inner.assert(pk).await
-        } else {
-            Err(Error::other("closed"))
+        let enc = self.inner.lock().unwrap().enc(pk, None)?;
+
+        {
+            let client = self.client.lock().await;
+            for enc in enc {
+                client.send(pk, &enc).await?;
+            }
         }
+
+        Ok(())
     }
 
     /// Send a message to a peer.
     pub async fn send(&self, pk: &PubKey, msg: &[u8]) -> Result<()> {
         const SBD_MAX: usize = 20_000;
         const SBD_HDR: usize = 32;
+        const SBD_P_HDR: usize = 1;
```

Comment:
here's that extra overhead byte

Reply:
Well hdr is a header, but what is a p?

Reply:
Fixed: https://github.com/holochain/sbd/pull/38/commits/c87d2cf38fac1c956469ccf8557f51b790dcb231